ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java

```diff
@@ -0,0 +1,252 @@
+  private final List<Long> partitionIds;
+  // Hashmap from partition id to the replicas
+  private final Map<Long, ReplicaId> partition2Replicas;
+  // generator to generate the correlation ids.
+  private final AtomicInteger correlationIdGenerator = new AtomicInteger(0);
+  // max result sets it can get from each query
+  private final int maxResults;
+  // shutdown flag
+  private volatile boolean shutdown = false;
+
+  /**
+   * RepairRequestsSender constructor
+   * @param requestChannel the channel to send requests and receive responses
+   * @param factory the {@link LocalNetworkClientFactory} used to generate the {@link LocalNetworkClient}
+   * @param clusterMap the cluster map
+   * @param nodeId this {@link DataNodeId}
+   * @param repairRequestsDb the database to get the RepairRequestRecord
+   * @param clusterParticipant {@link ClusterParticipant}
+   */
+  public RepairRequestsSender(RequestResponseChannel requestChannel, LocalNetworkClientFactory factory,
+      ClusterMap clusterMap, DataNodeId nodeId, RepairRequestsDb repairRequestsDb,
+      ClusterParticipant clusterParticipant) {
+    this.requestChannel = requestChannel;
+    this.client = factory.getNetworkClient();
+    this.clusterMap = clusterMap;
+    this.nodeId = nodeId;
+    this.clusterParticipant = clusterParticipant;
+
+    this.db = repairRequestsDb;
+    this.maxResults = db.getListMaxResults();
+
+    // LOCAL_CONSISTENCY_TODO. listen to the clustermap change and make the modification accordingly.
+    partitionIds = new ArrayList<>();
+    partition2Replicas = new HashMap<>();
+    for (ReplicaId replica : clusterMap.getReplicaIds(nodeId)) {
+      long partition = replica.getPartitionId().getId();
+      partitionIds.add(partition);
+      partition2Replicas.put(partition, replica);
+    }
+  }
+
+  public void run() {
+    List<RequestInfo> requestInfos = null;
+
+    while (!shutdown) {
+      try {
+        // sleep for some time
+        Thread.sleep(SLEEP_TIME_MS);
+
+        boolean hasMore = true;
+        boolean hasError = false;
+        // if some partitions have more requests, continue to handle them.
+        while (hasMore && !hasError) {
+          hasMore = false;
+          hasError = false;
+          // loop all the partitions. For each partition, we handle maxResults number of requests.
+          for (long partitionId : partitionIds) {
+            if (shutdown) {
+              return;
+            }
+            // get the repair requests for this partition
+            requestInfos = getAmbryRequest(partitionId);
+            if (requestInfos.isEmpty()) {
+              continue;
+            }
+            // this partition may have more requests.
+            if (requestInfos.size() >= maxResults) {
+              hasMore = true;
+            }
+
+            Set<Integer> requestsToDrop = new HashSet<>();
+            // send the repair requests to the handler and wait for the responses.
+            while (!requestInfos.isEmpty()) {
```

Comment:
this could just be a for-loop, like `for(RequestInfo requestInfo: requestInfos)`, with that, you don't have to remove the RequestInfo from the list.

Reply:
yes, the reason to use the get and remove is because of the memory release part I got confused. Will clean it up. Thanks,